src/diffusers/models/dit.py

```diff
@@ -0,0 +1,332 @@
+
+
+def to_2tuple(x: int):
+    if isinstance(x, collections.abc.Iterable) and not isinstance(x, str):
+        return x
+    return tuple(repeat(x, 2))
+
+
+class PatchEmbed(nn.Module):
+    """2D Image to Patch Embedding"""
+
+    def __init__(
+        self,
+        img_size=224,
+        patch_size=16,
+        in_chans=3,
+        embed_dim=768,
+        layer_norm=False,
+        flatten=True,
+        bias=True,
+    ):
+        super().__init__()
+        img_size = to_2tuple(img_size)
+        patch_size = to_2tuple(patch_size)
+        self.img_size = img_size
+        self.patch_size = patch_size
+
+        self.grid_size = (img_size[0] // patch_size[0], img_size[1] // patch_size[1])
+        self.num_patches = self.grid_size[0] * self.grid_size[1]
+        self.flatten = flatten
+
+        self.proj = nn.Conv2d(in_chans, embed_dim, kernel_size=patch_size, stride=patch_size, bias=bias)
+        self.norm = nn.LayerNorm(embed_dim, elementwise_affine=False, eps=1e-6) if layer_norm else nn.Identity()
+
+    def forward(self, latent):
+        _, _, height, width = latent.shape
+        if height != self.img_size[0] or width != self.img_size[1]:
+            ValueError(
+                f"Input image size ({height}x{width}) doesn't match model ({self.img_size[0]}x{self.img_size[1]})."
+            )
+
+        latent = self.proj(latent)
+        if self.flatten:
+            latent = latent.flatten(2).transpose(1, 2)  # BCHW -> BNC
+        latent = self.norm(latent)
+        return latent
+
+
+class Mlp(nn.Module):
+    """MLP as used in Vision Transformer, MLP-Mixer and related networks"""
+
+    def __init__(
+        self,
+        in_features,
+        hidden_features=None,
+        out_features=None,
+        activation_layer="gelu",
+        bias=True,
+        dropout_prob=0.0,
+    ):
+        super().__init__()
+        out_features = out_features or in_features
+        hidden_features = hidden_features or in_features
+
+        self.fc1 = nn.Linear(in_features, hidden_features, bias=bias)
+        if activation_layer == "gelu":
+            self.act = nn.GELU()
+        elif activation_layer == "gelu-approximate":
+            self.act = nn.GELU(approximate="tanh")
+
+        self.drop1 = nn.Dropout(dropout_prob)
+        self.fc2 = nn.Linear(hidden_features, out_features, bias=bias)
+        self.drop2 = nn.Dropout(dropout_prob)
+
+    def forward(self, latent):
+        latent = self.fc1(latent)
+        latent = self.act(latent)
+        latent = self.drop1(latent)
+        latent = self.fc2(latent)
+        latent = self.drop2(latent)
+        return latent
+
+
+class DiTBlock(nn.Module):
+    """
+    A DiT block with adaptive layer norm zero (adaLN-Zero) conditioning.
+    """
+
+    def __init__(self, hidden_size, num_heads, mlp_ratio=4.0):
+        super().__init__()
+        self.norm1 = nn.LayerNorm(hidden_size, elementwise_affine=False, eps=1e-6)
+        self.attn = CrossAttention(
```

Comment:
Very nice that we can re-use the cross attention class here! 

Could we maybe re-use the complete https://github.com/huggingface/diffusers/blob/8c14ca3d434c1ff5b118f0b18e1db20fea73c51f/src/diffusers/models/transformer_2d.py#L40 model here ? 

We already have an `if statement` [here](https://github.com/huggingface/diffusers/blob/8c14ca3d434c1ff5b118f0b18e1db20fea73c51f/src/diffusers/models/attention.py#L248) with ada layer norm: https://github.com/huggingface/diffusers/blob/8c14ca3d434c1ff5b118f0b18e1db20fea73c51f/src/diffusers/models/attention.py#L386 which looks very similar / identical